### PR TITLE
socket enricher: Use single instance for multiple gadgets

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
Check the details in https://github.com/inspektor-gadget/inspektor-gadget/issues/1695.

Fixes https://github.com/inspektor-gadget/inspektor-gadget/issues/1695

### CODE REVIEW

File: pkg/operators/socketenricher/socketenricher.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+type SocketEnricherInstance struct {
+	gadgetCtx      operators.GadgetContext
+	manager        *SocketEnricher
+	gadgetInstance any
+}
+
+func (i *SocketEnricherInstance) Name() string {
+	return "SocketEnrichreInstance"
```
Comment (alban) [inline pkg/operators/socketenricher/socketenricher.go:114]: s/SocketEnrichreInstance/SocketEnricherInstance

File: pkg/gadgets/run/tracer/tracer.go [hunk 2]
```diff
@@ -894,3 +886,7 @@ func (t *Tracer) SetEventHandlerArray(handler any) {
 	}
 	t.eventArrayCallback = nh
 }
+
+func (t *Tracer) SetSocketEnricherMap(m *ebpf.Map) {
+	t.socketEnricherMap = m
```
Comment (alban) [inline pkg/gadgets/run/tracer/tracer.go:891]: Does it mean that image-based gadgets would still inconditionally create the socket enricher, even if they don't use it?

**Review by alban — COMMENTED** [review]
Thanks!

File: pkg/gadgets/run/tracer/tracer.go (hunk 2 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/tracer/tracer.go:891]: You're right. It's still a pending issue we have with other operators as well (https://github.com/inspektor-gadget/inspektor-gadget/issues/2167), I gave it a try on https://github.com/inspektor-gadget/inspektor-gadget/pull/2168.

**Comment by mauriciovasquezbernal** [comment]
There's something going on with the tests on this PR. Holding until I have the chance to investigate more.

**Comment by burak-ok** [comment]
> There's something going on with the tests on this PR. Holding until I have the chance to investigate more.

At least for `TestProfileCpu` the KubeInventoryCache crashed the gadget. After rebase + https://github.com/inspektor-gadget/inspektor-gadget/pull/2354 at least one test should pass again.
That crash seems to be a very rare race condition

**Comment by mauriciovasquezbernal** [comment]
@burak-ok does it have any relation with https://github.com/inspektor-gadget/inspektor-gadget/pull/2347? Because that PR is not included in this branch yet.

**Comment by burak-ok** [comment]
> @burak-ok does it have any relation with #2347? Because that PR is not included in this branch yet.

No, that rare race condition appears to have existed until that branch. With #2347 I think it got worse

File: pkg/gadget-collection/gadgets/trace/network/gadget.go [hunk 3]
```diff
@@ -127,13 +129,21 @@ func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
...
 		t.kubeNameInst.PreGadgetRun()
 	}
 
+	socketEnricherOp := operators.GetRaw(socketenricher.OperatorName).(*socketenricher.SocketEnricher)
+	socketEnricherOp.Init(nil)
+	t.socketEnricherInst, err = socketEnricherOp.Instantiate(nil, t.tracer, nil)
+	if err == nil {
+		t.socketEnricherInst.PreGadgetRun()
+	}
+
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadget-collection/gadgets/trace/network/gadget.go:146]: This whole file is legacy and all this operators-related stuff there is a big hack, I hope we can get rid of it very soon.

File: pkg/gadgets/trace/network/tracer/tracer.go [hunk 4]
```diff
@@ -48,12 +48,18 @@ func NewTracer() (_ *Tracer, err error) {
 		return nil, fmt.Errorf("installing tracer: %w", err)
 	}
 
+	return t, nil
+}
+
+// RunWorkaround is used by pkg/gadget-collection/gadgets/trace/network/gadget.go to run the gadget
+// after calling NewTracer(). It should be removed one the above file is removed.
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/gadgets/trace/network/tracer/tracer.go:55]: I was referring to pkg/gadget-collection/gadgets/trace/network/gadget.go. Change the sentence to be more clear.

File: pkg/socketenricher/tracer.go [hunk 5]
```diff
@@ -243,4 +243,5 @@ func (se *SocketEnricher) Close() {
 	}
 	se.links = nil
 	se.objs.Close()
+	se.objsIter.Close()
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/socketenricher/tracer.go:246]: I think it's. However I can't remember if there was any user visible impact of this on main (I opened this PR a long time ago).

File: pkg/operators/socketenricher/socketenricher.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	return nil
+}
+
+func (i *SocketEnricherInstance) PostGadgetRun() error {
+	i.manager.mu.Lock()
+	defer i.manager.mu.Unlock()
+
+	i.manager.refCount--
+	if i.manager.refCount == 0 {
+		i.manager.socketEnricher.Close()
```
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/operators/socketenricher/socketenricher.go:147]: We could but I think the semantics are not the same. `(s *SocketEnricher) Close()` should destroy everything related to the parent, while `(i *SocketEnricherInstance) PostGadgetRun()` is only for this instance. They are exactly the same now, but they could be different later on.

**Pull request merged**